### PR TITLE
8344458: [11u] Add initial support for building with XCode 14

### PR DESCRIPTION
**Pull request by vieiro**
An implementation of [JDK-8344458](https://bugs.openjdk.org/browse/JDK-8344458) that adds conditional support for building on `macos` with `XCode 14` while keeping compatibility with previous `XCode` versions.

The PR is separated in three commits for easier review:

- First commit adds a new `--enable-xcode14` configuration flag (which is currently disabled in GHA) and two additional variables:
    - `CFLAGS_XCODE14_DEPR_DECLARATIONS` (empty on current XCode versions, set to and `-Wno-deprecated-declarations` when `--enable-xcode14`is used)
    - And `CFLAGS_XCODE14_DEPR_NON_PROTOTYPE` (empty on current XCode versions, set to `-Wno-deprecated-non-prototype` when `--enable-xcode14` is used).
- The second commit applies `CFLAGS_XCODE14_DEPR_DECLARATIONS` to those parts of the codebase that use the deprecated `sprintf` function (i.e., avoiding `sprintf` usage errors in XCode 14).
- The third commit applies `CFLAGS_XCODE14_DEPR_NON_PROTOTYPE` to those parts of AWT lib that generate a compilation error (i.e., avoiding the `passing arguments to a function without prototype`).

Since the new flag `--enable-xcode14` is not set **the build should run exactly the same on the current XCode versions** and, consequently, the GitHub checks should pass on all platforms, including the current `macos-12` & `XCode 13.4.1`.

---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [ ] [JDK-8344458](https://bugs.openjdk.org/browse/JDK-8344458) needs maintainer approval
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8344458](https://bugs.openjdk.org/browse/JDK-8344458): [11u] Add initial support for building with XCode 14 (**Enhancement** - P4)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk11u-dev.git pull/2966/head:pull/2966` \
`$ git checkout pull/2966`

Update a local copy of the PR: \
`$ git checkout pull/2966` \
`$ git pull https://git.openjdk.org/jdk11u-dev.git pull/2966/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 2966`

View PR using the GUI difftool: \
`$ git pr show -t 2966`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk11u-dev/pull/2966.diff">https://git.openjdk.org/jdk11u-dev/pull/2966.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk11u-dev/pull/2966#issuecomment-2484248418)
</details>

### CODE REVIEW

**Comment by vieiro** [comment]
NOTE: [The GHA `macos-12`  runners have been shut-down](https://github.com/actions/runner-images/issues/10721)  between November 18, 14:00 UTC and November 19, 00:00 UTC and are currently unavailable. The tests will be re-run when again available.

File: doc/building.html [hunk 1]
```diff
@@ -314,6 +314,7 @@ <h3 id="apple-xcode">Apple Xcode</h3>
 <li>Use configure option <code>--with-xcode-path</code>, e.g. <code>configure --with-xcode-path=/Applications/Xcode13.1.app</code> This allows using a specific Xcode version for an OpenJDK build, independently of the active Xcode version by <code>xcode-select</code>.</li>
 </ul>
 <p>If you have recently (inadvertently) updated your OS and/or Xcode version, and the JDK can no longer be built, please see the section on <a href="#problems-with-the-build-environment">Problems with the Build Environment</a>, and <a href="#getting-help">Getting Help</a> to find out if there are any recent, non-merged patches available for this update.</p>
+<p>Experimental support for XCode 14 can be enabled using the <code>--enable-xcode14</code> configuration flag.</p>
```
Comment (jerboaa) [inline doc/building.html:317]: Please remove this. We don't usually update `doc/building.html` for every build change.

File: doc/building.md [hunk 2]
```diff
@@ -355,6 +355,8 @@ Build Environment](#problems-with-the-build-environment), and [Getting
 Help](#getting-help) to find out if there are any recent, non-merged patches
 available for this update.
 
+Experimental support for XCode 14 can be enabled using the `--enable-xcode14` configuration flag.
+
```
Comment (jerboaa) [inline doc/building.md:359]: Same here.

**Review by jerboaa — COMMENTED** [review]
Seems OK to me. I'm not sure we need the doc fixes.

File: doc/building.html (hunk 1 above)
Comment (vieiro) [reply doc/building.html]: Removed in last commit

File: doc/building.md (hunk 2 above)
Comment (vieiro) [reply doc/building.md]: Removed in last commit.

**Comment by vieiro** [comment]
> [...]
> If these flags are needed for builds with XCode 14, we should detect that this newer compiler is being used and add them automatically. In fact, ideally, you should not depend on the version at all but test the compiler with the flag and add it if it works. Take a look at `FLAGS_SETUP_GCC6_COMPILER_FLAGS` also in `flags-cflags.m4`
> [...]

Yep, I came to the same conclusion in the updates_dev mailing list after pushing. The flag was useful to try to compile on XCode 13 and XCode 14 during experiments, but detecting the compiler is probably the way to go.

**Comment by vieiro** [comment]
> [...]
> Is `--disable-warnings-as-errors` not sufficient to make the GHA build work on MacOS 13? It's not an ideal solution but it's a less disruptive way to keep a working MacOS build in the short term.

I'll give it a try. Yep, it's not ideal (we may not be able to run the tests) but it's indeed less disruptive.

**Comment by vieiro** [comment]
Inspired by this comment

> Is `--disable-warnings-as-errors` not sufficient to make the GHA build work on MacOS 13? It's not an ideal solution but it's a less disruptive way to keep a working MacOS build in the short term.

I thought we could augment the `DISABLED_WARNINGS` on the macosx/clang 14.X.Y combo with `unused-parameter unused deprecated-declarations deprecated-non-prototype`.

This change has no effect in previous clang versions, so builds and tests on these previous versions are unaffected.

**Comment by vieiro** [comment]
Closed in favor of [JDK-8343786](https://bugs.openjdk.org/browse/JDK-8343786) in #2967

**Comment by gnu-andrew** [comment]
I don't think #2967 really replaces this as the issue is still there. We are just working around it in GHA.

**Pull request closed without merging**